src/Workspaces/CSharp/Portable/Formatting/Rules/BaseFormattingRule.cs

```diff
@@ -166,5 +167,82 @@ protected AdjustSpacesOperation CreateAdjustSpacesOperation(int space, AdjustSpa
         {
             return FormattingOperations.CreateAdjustSpacesOperation(space, option);
         }
+
+        protected void AddBraceSuppressOperations(List<SuppressOperation> list, SyntaxNode node, SyntaxToken lastToken)
+        {
+            var bracePair = node.GetBracePair();
+            if (!bracePair.IsValidBracePair())
+            {
+                return;
+            }
+
+            var firstTokenOfNode = node.GetFirstToken(includeZeroWidth: true);
+
+            if (node.IsLambdaBodyBlock())
+            {
+                // include lambda itself.
+                firstTokenOfNode = node.Parent.GetFirstToken(includeZeroWidth: true);
+            }
+
+            // We may think we have a complete set of braces, but that may not actually be the case
+            // due incomplete code.  i.e. we have something like:
+            //
+            // class C
+            // {
+            //      int Blah {
+            //          get { return blah
+            // }
+            //
+            // In this case the parse will think that the get-accessor is actually on two lines 
+            // (because it will consume the close curly that more accurately belongs to the class.
+            //
+            // Now there are different behaviors we want depending on what the user is doing 
+            // and what we are formatting.  For example, if the user hits semicolon at the end of
+            // "blah", then we want to keep the accessor on a single line.  In this scenario we
+            // effectively want to ignore the following close curly as it may not be important to
+            // this construct in the mind of the user. 
+            //
+            // However, say the user hits semicolon, then hits enter, then types a close curly.
+            // In this scenario we woudl actually want the get-accessor to be formatted over multiple 
+            // lines.  The difference here is that because the user just hit close-curly here we can 
+            // consider it as being part of the closest construct and we can consider its placement
+            // when deciding if the construct is on a single line.
+
+            var endBrace = bracePair.Item2;
+            if (lastToken.Kind() != SyntaxKind.CloseBraceToken && 
```

Comment:
Here's the 'meat' of the review where we actually look at what token the user inputted, and we also try to determine if we're in a situation where we should actually treat the block as if it had a missing close curly.